File: bit-manipulation/single-number-deus/SingleNumberDeus.java
// https://leetcode.com/problems/single-number-ii/?envType=study-plan-v2&envId=top-interview-150

class Solution {
    public int singleNumber(int[] nums) {

        int result = 0;

        for(int i=0;i<32;i++) {

            int bitSum = 0;

            // get the ith bit from each number
            // then add them up
            for(int num: nums) {
                // String bits = String.format("%32s",
                //         Integer.toBinaryString(num)).replace(' ', '0');
                // System.out.println("the 32 bit representation of number: " + num + " is: " + bits);

                // how to get the ith bit in a number
                // 1. Shift the desired bit to the least significant position -> this is the rightmost bit in the
                // binary representation of a number - example: for binary 001 -> 1 is the least significant bit
                // this is acheived by the num >> i command -> you're essentially shifting the bits in the number num by
                // i positions, in case the number is positive the new bits that are added to the left are 0, otherwise they are
                // 1
                // 2. Use the AND operator with 1 (which is 0...001) to isolate that bit
                int bit = (num >> i) & 1;

                // System.out.println("ith bit for this number is: " + String.format("%32s",
                //         Integer.toBinaryString(bit)).replace(' ', '0'));

                bitSum = bitSum + bit;
            }

            int finalBit = bitSum % 3;

            // set the ith bit as finalBit
            // essentially shift finalBit left by i positions, then add it up to the original result
            result = result | (finalBit << i);
        }

        return result;
    }
}

File: bit-manipulation/single-number-deus/single_number_deus.py
# https://leetcode.com/problems/single-number-ii/?envType=study-plan-v2&envId=top-interview-150

# Time complexity: O(n) where n is the list of numbers
# Space complexity: O(1)

class Solution:
    def singleNumber(self, nums: List[int]) -> int:

        result = 0

        for index in range (32):

            bit_sum = 0
            
            for num in nums:

                # print("bin representation of original number: ", num)
                # print(bin(num)[2:])

                # print("number after right shifting", index, " bits: ")
                right_shifted_number = num >> index
                # print(bin(right_shifted_number)[2:])

                # how do I get the ith bit in a number
                # fairly simple you "AND" everything with one - dang that's amazing

                # print("bit", index , "from the number:", num , " is:")
                bit = right_shifted_number & 1
                # print(bin(bit)[2:])

                bit_sum = bit_sum + bit

            result_bit = bit_sum % 3
            result_bit = result_bit << index
            result = result | result_bit
            # print("This is the bit that should be returned in the result: ", result_bit)

        # Convert from unsigned 32-bit to signed integer.
        # If the highest bit (sign bit) is set, the number is actually negative.
        # Subtracting 2^32 converts it to the correct negative value.
        # Example: 4294967292 → -4
        if result >= 2**31:
            result -= 2**32
        
        return result